### DISCUSSION TITLE
v0.1.0

### DISCUSSION THREAD

**Discussion by rdbende** [tukaan/tukaan#83] [Announcements]
### What's Changed
* Rewrite quasi everything by @rdbende in https://github.com/tukaan/tukaan/pull/82

**Full Changelog**: https://github.com/tukaan/tukaan/compare/v0.0.1.dev1...v0.1.0
---

\* There's no multiline textbox widget at the moment, however the `Entry` widget was renamed to `TextBox`, this might be confusing 😕
The multiline text widget will return, but in [a revised form](http://scidb.sourceforge.net/tk/revised-text-widget.html), at least until it is merged into the `Tcl/Tk` core

<hr /><em>This discussion was created from the release <a href='https://github.com/tukaan/tukaan/releases/tag/v0.1.0'>v0.1.0</a>.</em>

**Comment by Moosems** [comment]
> * There's no multiline textbox widget at the moment, however the Entry widget was renamed to TextBox, this might be confusing 😕
The multiline text widget will return, but in [a revised form](http://scidb.sourceforge.net/tk/revised-text-widget.html), at least until it is merged into the Tcl/Tk core

Why? Just curious. Should I update the docs to reflect this?

> Not backwards-compatible

What do you mean? Does my old code not work?

Side: <br> Was this rewritten for the layout change or something else? Just curious. Sorry for all the questions I'm just very confused

**Reply by sumeshir26** [reply]
I think the code was very cluttered so rdbende rewrote it

**Reply by rdbende** [reply]
I totally understand that you're confused.
I know, this change is ridiculous (looking at the diff, even I think it's insane). I wanted to rewrite it, because the old code was very cluttered and also buggy sometimes and overwriting the `__setattr__` and `__getattr__` methods had some serious limitations andoften led to a `RecursionError`, and I had to use incredible hacks to get around it.

As for the `TextBox` widget:
I knew, that Tk's text widget is really buggy, often ran into these bugs, but when I found [this](http://scidb.sourceforge.net/tk/revised-text-widget.html) and found out how buggy it really is, I figured it would be nice to use this revised implementation instead in Tukaan, at least until it gets merged into the Tcl/Tk core: [TIP 466](https://core.tcl-lang.org/tips/doc/trunk/tip/466.md))

I renamed the `Entry` widget to `TextBox` to be more consistent with the `ComboBox` and `SpinBox` widgets, and the `TextBox` widget to `TextView`.

> What do you mean? Does my old code not work?

It might not.
The project is still in beta state, so I don't recommend anyone to write a program with it, since the code may contain backwards incompatible changes.

In this patch I did the following:
- renamed a lot of things, widgets, arguments, methods, properties
- rewrote the layouting (needed some change to the widgets)
- rewrote some code logic
- modernized amd simplified the code
- replaced the `__(s|g)etattr__` with properies
- fixed a lotta bugs

**Reply by Moosems** [reply]
@rdbende thank you for the update and explanation. After hearing your explanation it makes much more sense and am grateful for your work. I look forward to using this new version!

**Reply by Moosems** [reply]
> The project is still in beta state, so I don't recommend anyone to write a program with it, since the code may contain backwards incompatible changes.

Isn't that even more reason to? It allows us to find minuscule issues that we wouldn't find otherwise

**Reply by rdbende** [reply]
Actually yes. On the one hand it's a good thing that people try out the library, and test the code, it really help the development. But on the other hand, the project is still unstable, and any vhange can brake your program, so that's why I don't recommend people to depend on it.

**Reply by Moosems** [reply]
So create programs and use it but don't depend on it

**Comment by Moosems** [comment]
ISSUE:

Mac build doesn't work. 

Error code:
```
File "/Users/Moosems/Desktop/JDE/test.py", line 881, in <module>
    app = App("Test")
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/tukaan/app.py", line 41, in __init__
    Serif.load()
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/tukaan/Serif/__init__.py", line 16, in load
    Tcl.call(None, "load", Path(__file__).parent / "pkg" / libfilename, "Serif")
  File "/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/tukaan/_tcl.py", line 168, in call
    raise TclError(msg) from None
tukaan.exceptions.TclError: dlopen(/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/tukaan/Serif/pkg/serif_mac_x64.dylib, 6): image not found
```

**Reply by Moosems** [reply]
Can we upload a quick fix?

**Reply by rdbende** [reply]
Uhh, indeed. The Mac binary for Serif doesn't exists yet. I'll fix it ASAP.

**Reply by Moosems** [reply]
Thanks. Sorry about that. I also am wondering when we can merge the tooltip into the main branch. What should I do to prepare it for that?

**Reply by rdbende** [reply]
I need to push a quick fix (the `.app` problem, I mentioned), and then the tooltip should be done.

**Reply by rdbende** [reply]
Mac issues fixed in `v0.1.1` 🎉

**Reply by Moosems** [reply]
Thank you so much!!
